src/main/scala/org/broadinstitute/hail/keytable/KeyTable.scala

```diff
@@ -0,0 +1,397 @@
+    require(keySignature == other.keySignature)
+
+    val (newValueSignature, merger) = valueSignature.merge(other.valueSignature)
+    val newRDD = rdd.fullOuterJoin(other.rdd).map { case (k, (vl, vr)) => (k, merger(vl.orNull, vr.orNull)) }
+
+    KeyTable(newRDD, keySignature, newValueSignature)
+  }
+
+  def innerJoin(other: KeyTable): KeyTable = {
+    require(keySignature == other.keySignature)
+
+    val (newValueSignature, merger) = valueSignature.merge(other.valueSignature)
+    val newRDD = rdd.join(other.rdd).map { case (k, (vl, vr)) => (k, merger(vl, vr)) }
+
+    KeyTable(newRDD, keySignature, newValueSignature)
+  }
+
+  def forall(code: String): Boolean = {
+    val ec = EvalContext(fields.map(f => (f.name, f.`type`)): _*)
+    val nKeysLocal = nKeys
+    val nValuesLocal = nValues
+
+    val f: () => Option[Boolean] = Parser.parse[Boolean](code, ec, TBoolean)
+
+    rdd.forall { case (k, v) =>
+      KeyTable.setEvalContext(ec, k, v, nKeysLocal, nValuesLocal)
+      f().getOrElse(false)
+    }
+  }
+
+  def exists(code: String): Boolean = {
+    val ec = EvalContext(fields.map(f => (f.name, f.`type`)): _*)
+    val nKeysLocal = nKeys
+    val nValuesLocal = nValues
+
+    val f: () => Option[Boolean] = Parser.parse[Boolean](code, ec, TBoolean)
+
+    rdd.exists { case (k, v) =>
+      KeyTable.setEvalContext(ec, k, v, nKeysLocal, nValuesLocal)
+      f().getOrElse(false)
+    }
+  }
+
+  def export(sc: SparkContext, output: String, typesFile: String) = {
+    val hConf = sc.hadoopConfiguration
+
+    val ec = EvalContext(fields.map(fd => (fd.name, fd.`type`)): _*)
+
+    val (header, types, f) = Parser.parseNamedArgs(fieldNames.map(n => n + " = " + n).mkString(","), ec)
+
+    Option(typesFile).foreach { file =>
+      val typeInfo = header
+        .getOrElse(types.indices.map(i => s"_$i").toArray)
+        .zip(types)
+
+      KeyTable.exportTypes(file, hConf, typeInfo)
+    }
+
+    hConf.delete(output, recursive = true)
+
+    val nKeysLocal = nKeys
+    val nValuesLocal = nValues
+
+    rdd
+      .mapPartitions { it =>
+        val sb = new StringBuilder()
+        it.map { case (k, v) =>
+          sb.clear()
+          KeyTable.setEvalContext(ec, k, v, nKeysLocal, nValuesLocal)
+          f().foreachBetween(x => sb.append(x))(sb += '\t')
+          sb.result()
+        }
+      }.writeTable(output, header.map(_.mkString("\t")))
+  }
+
+  def aggregate(keyCond: String, aggCond: String): KeyTable = {
+
+    val aggregationEC = EvalContext(fields.map(fd => (fd.name, fd.`type`)): _*)
+    val ec = EvalContext(fields.zipWithIndex.map { case (fd, i) => (fd.name, (-1, KeyTableAggregable(aggregationEC, fd.`type`, i))) }.toMap)
+
+    val (keyNameParseTypes, keyF) =
+      if (keyCond != null)
+        Parser.parseAnnotationArgs(keyCond, aggregationEC, None)
+      else
+        (Array.empty[(List[String], Type)], Array.empty[() => Any])
+
+    val (aggNameParseTypes, aggF) =
+      if (aggCond != null)
```

Comment:
Again with nulls here.  I gather this is to interface with python...  Is this better than using empty string?  What happens if you pass empty string?